### PR TITLE
Extension everything initial PR

### PR DESCRIPTION
**Pull request by khyperia**
This is the initial code for Extension Everything. This PR is more of an FYI of current status, and all future changes will go through code review.

The state of the code is still pretty messy, and there are a lot of outstanding issues that I haven't gotten around to fixing yet - a code review might not be the most useful thing, as there's still a lot of clean-up work I know of (and listing remaining work items on a public document is itself a remaining work item :smiley:)

Ping @dotnet/roslyn-compiler as an FYI

### CODE REVIEW

**Comment by khyperia** [comment]
retest perf_correctness_prtest please

**Comment by jaredpar** [comment]
@khyperia it looks like you are hitting the perf correctness issues because of self host problems.  For now we should make this job trigger only until we fix the self host bug.  Go to netci.groovy and change this line to `true`

https://github.com/dotnet/roslyn/blob/master/netci.groovy#L178

File: src/Compilers/CSharp/Portable/Symbols/NamespaceSymbol.cs [hunk 1]
```diff
@@ -315,48 +315,47 @@ internal NamespaceSymbol GetNestedNamespace(NameSyntax name)
...
+        /// Add all extension members in this namespace to the given list. If name or arity
         /// or both are provided, only those extension methods that match are included.
         /// </summary>
-        /// <param name="methods">Methods list</param>
+        /// <param name="members">Members list</param>
         /// <param name="nameOpt">Optional method name</param>
         /// <param name="arity">Method arity</param>
         /// <param name="options">Lookup options</param>
-        internal virtual void GetExtensionMethods(ArrayBuilder<MethodSymbol> methods, string nameOpt, int arity, LookupOptions options)
+        internal virtual void GetExtensionMembers(ArrayBuilder<Symbol> members, string nameOpt, int arity, LookupOptions options)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/NamespaceSymbol.cs]: In the description, it is probably worth explicitly talking about reduced form of methods returned by this method.

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingMethodSymbol.cs [hunk 2]
```diff
@@ -86,6 +86,15 @@ public override MethodSymbol UnderlyingMethod
             }
         }
 
+        public override TypeSymbol ReceiverType
+        {
+            get
+            {
+                var receiver = _underlyingMethod.ReceiverType;
+                return this.RetargetingTranslator.Retarget(receiver, RetargetOptions.RetargetPrimitiveTypesByTypeCode);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingMethodSymbol.cs]: Is RetargetPrimitiveTypesByTypeCode the right option to use? Is this type going to be encoded in metadata by using type codes?

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamespaceSymbol.cs [hunk 3]
```diff
@@ -216,15 +216,27 @@ internal override NamedTypeSymbol LookupMetadataType(ref MetadataTypeName typeNa
...
         {
-            var underlyingMethods = ArrayBuilder<MethodSymbol>.GetInstance();
-            _underlyingNamespace.GetExtensionMethods(underlyingMethods, nameOpt, arity, options);
-            foreach (var underlyingMethod in underlyingMethods)
+            var underlyingMembers = ArrayBuilder<Symbol>.GetInstance();
+            _underlyingNamespace.GetExtensionMembers(underlyingMembers, nameOpt, arity, options);
+            foreach (var underlyingMember in underlyingMembers)
             {
-                methods.Add(this.RetargetingTranslator.Retarget(underlyingMethod));
+                if (underlyingMember.Kind == SymbolKind.Method)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamespaceSymbol.cs]: Please assert that underlyingMember.IsDefinition before this 'if'

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamespaceSymbol.cs [hunk 4]
```diff
@@ -216,15 +216,27 @@ internal override NamedTypeSymbol LookupMetadataType(ref MetadataTypeName typeNa
...
+                {
+                    var underlyingMethod = (MethodSymbol)underlyingMember;
+                    if (underlyingMethod.MethodKind == MethodKind.ReducedExtension)
+                    {
+                        var original = underlyingMethod.UnreduceExtensionMethod();
+                        var retargeted = this.RetargetingTranslator.Retarget(original);
+                        var reduced = retargeted.ReduceExtensionMethod();
+                        members.Add(reduced);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamespaceSymbol.cs]: Please assert that reduced.IsDefinition. Also ensure that it is not null.

**Pull request merged**